modules/table/src/main/java/org/apache/ignite/internal/table/distributed/replicator/PartitionReplicaListener.java

```diff
@@ -486,11 +487,23 @@ private CompletableFuture<BinaryRow> processReadOnlySingleEntryAction(ReadOnlySi
                     format("Unknown single request [actionType={}]", request.requestType()));
         }
 
-        CompletableFuture<Void> safeReadFuture = isPrimary ? completedFuture(null) : safeTime.waitFor(request.readTimestamp());
+        CompletableFuture<Void> safeReadFuture = isPrimaryInTimestamp(isPrimary, readTimestamp) ? completedFuture(null)
+                : safeTime.waitFor(request.readTimestamp());
 
         return safeReadFuture.thenCompose(unused -> resolveRowByPkForReadOnly(searchRow, readTimestamp));
     }
 
+    /**
+     * Checks that the node is primary in a passed timestamp.
+     *
+     * @param isPrimary True is the node is primary, false otherwise.
+     * @param timestamp Timestamp to check.
+     * @return True if the timestamp is already come and node in primary, otherwise false.
```

Comment:
```suggestion
     * @return True if the timestamp is already passed in the reference system of the current node and node is primary, false otherwise.
```